### PR TITLE
ci: test container before push

### PR DESCRIPTION
**Pull request by huxuan**
Now devcotnainer test will also run when related changes are made when sending a pull reqeust.
Both dev and prod containers will only be pushed when passing tests.

### CODE REVIEW

File: .github/workflows/devcontainer.yml [hunk 1]
```diff
@@ -18,26 +36,38 @@ jobs:
           docker buildx build . \
```
Comment (msclock) [inline .github/workflows/devcontainer.yml:36]:
```suggestion
          echo "Building a development image"
          docker buildx build . \
```
maybe we can add some annotations for run steps.

File: .github/workflows/devcontainer.yml [hunk 2]
```diff
@@ -18,26 +36,38 @@ jobs:
...
             --build-arg PYTHON_VERSION=${{ matrix.python-version }} \
             --cache-from type=registry,ref=ghcr.io/${{ github.repository }}/dev-cache:py${{ matrix.python-version }} \
+            --file .devcontainer/Dockerfile \
+            --load \
+            --tag ghcr.io/${{ github.repository }}/dev:py${{ matrix.python-version }} \
+            --target dev
+      - run: |
+          docker run --rm \
```
Comment (msclock) [inline .github/workflows/devcontainer.yml:44]:
```suggestion
          echo "Testing a development image"
          docker run --rm \
```

File: .github/workflows/devcontainer.yml [hunk 3]
```diff
@@ -18,26 +36,38 @@ jobs:
...
+            -e CI=true \
+            -v ${PWD}:/workspace \
+            ghcr.io/${{ github.repository }}/dev:py${{ matrix.python-version }} \
+            make dev lint test doc build
+      - run: |
+          docker buildx build . \
```
Comment (msclock) [inline .github/workflows/devcontainer.yml:50]:
```suggestion
          echo "Building a production image."
          docker buildx build . \
```

File: .github/workflows/devcontainer.yml (hunk 1 above)
Comment (huxuan) [reply .github/workflows/devcontainer.yml:36]: Good point. I created an issue #501 to add proper names, maybe we can do it there?

Comment (msclock) [reply .github/workflows/devcontainer.yml:36]: Got it.

**Review by msclock — APPROVED** [review]

**Pull request merged**